### ISSUE TITLE
Is something wrong I am doing?

### ISSUE THREAD

**Issue by chaitanya-smartrotamac** [curiores/SerialAnalyzer#4]
HI @curiores 
This is a wonderful program and a great idea.
I like your program very much.

But I have a problem

![image](https://user-images.githubusercontent.com/73290321/207323247-b04e22c8-b1e9-4a18-945b-64e04fbc39b2.png)

here you see 3 frequency components

As I coded Arduino as this
xn = sin(2*PI*2*t) + 0.2*sin(2*PI*50*t) + 0.8*sin(2*PI*30*t) + 0.5*sin(2*PI*100*t);

But up to 90Hz whatever frequency I used in the formula it is showing correct but if it crosses 90 degrees it kind of warp

See here I used  xn = sin(2*PI*2*t) + 0.2*sin(2*PI*150*t);
![image](https://user-images.githubusercontent.com/73290321/207324580-eb4b3332-6663-41b9-880a-0682a0822e23.png)

but when I used 120 it not worked
![image](https://user-images.githubusercontent.com/73290321/207326418-29584a90-70f6-4236-b7a7-f70fc63edb06.png)

when I use 100 it worked
![image](https://user-images.githubusercontent.com/73290321/207326692-81963e1f-95c5-481e-b8f2-7cc4fd86ddc6.png)

Shouldnt I use frequencies like that?
Am I missing some very basics of trigonometry?

**Comment by curiores** [comment]
There's nothing wrong with those frequencies, but the sample rate you're using can't support them. If your sample rate is too low, you won't pick up the correct frequencies, you'll pick up aliasing (like when we see a car wheel spinning backward). On an Arduino, your sample rate is limited by how quickly your loop function runs. If you can get your sample rate higher, there won't be an issue. The easiest way to do that is crank up the baud rate. I think the maximum for Arduino is 115200.

**Comment by chaitanya-smartrotamac** [comment]
Thank you @curiores for two things.
1. Quick response.
2. I Really always admire your explanation, Very crisp and straightforward. Your clarified my doubt with just one word "aliasing" 

Thank you.

So I will change the slow Arduino with a powerful controller to generate proper time sampling to get all frequencies measured.
can software handle 2Mbps? I am sure it does... but will check.

Forgot to mention : 
just removing the 3 times condition in code and achieved frequency analysis up to 150Hz, That is quite enough with a simple setup like Arduino
xn = sin(2*PI*2*t) + 0.2*sin(2*PI*10*t) + 0.4*sin(2*PI*20*t) + 0.6*sin(2*PI*30*t) + 0.8*sin(2*PI*50*t)+ 0.2 * sin(2*PI*80*t)+ 0.5* sin(2*PI*120*t) + 0.8 * sin(2*PI*150*t)

![image](https://user-images.githubusercontent.com/73290321/207515618-175e709d-9ec5-4044-91d3-b05ea6df8b0f.png)
.

removing delay(1) also reduced a millisecond of time and the sampling rate increased and I can see no problem up to 250Hz.
and that's great.

Thank you again.
_You earned a subscriber_

**Comment by curiores** [comment]
Hi @chaitanya-smartrotamac,

I didn't test the maximum data stream the software can handle. It's written in JS libraries so to be fair it's not exactly the most performant software. If I recall correctly I added some options in there that can slow the visual updates without actually slowing the data stream. The main performance issues I noticed was when a lot of graphical stuff was updated too often.

Looks good! Yeah if you put a delay in there it's going to restrict the sample rate.

Glad it helped. You asked the question with clarity and specificity, so you earned the quick response.
